### PR TITLE
The grand refactoring

### PR DESCRIPTION
**Pull request by pichfl**
Please do not merge yet, it's def. incomplete and I created several things that need to be fixed before this can be merged. Feel free to add comments anyway. 

I can't promise I'll be able to separate all the changes into smaller PRs so I'm not trying right now, but I can do so if there is no other way. This replaces #62

### CODE REVIEW

**Comment by GianniCarlo** [comment]
so far it looks great! I'll delve a little bit deeper later and run it locally to see how it goes 👍

File: Audiobook Player/Extensions/Extensions.swift [hunk 1]
```diff
@@ -52,5 +51,57 @@ extension Notification.Name {
         public static let bookPaused = Notification.Name(rawValue: "com.tortugapower.audiobookplayer.book.pause")
         public static let bookEnd = Notification.Name(rawValue: "com.tortugapower.audiobookplayer.book.end")
         public static let bookChange = Notification.Name(rawValue: "com.tortugapower.audiobookplayer.book.change")
+        public static let bookPlaying = Notification.Name(rawValue: "com.tortugapower.audiobookplayer.book.playback")
```
Comment (GianniCarlo) [inline Audiobook Player/Extensions/Extensions.swift]: we need to delete `updateTimer` then

File: Audiobook Player/Models/Book.swift [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+                UserDefaults.standard.set(nil, forKey: UserDefaultsConstants.lastPauseTime+"_\(self.identifier)")
+            }
+        }
+
+        return lastPlayedPosition
+    }
+
+    var percentCompleted: Int = 0
+    var percentCompletedString: String = "0%"
```
Comment (GianniCarlo) [inline Audiobook Player/Models/Book.swift]: we could have this one be a computed property too based on `percentCompleted`

File: Audiobook Player/Models/Book.swift [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    // MARK: Chapters
+
+    var chapters: [Chapter] = [] {
+        didSet {
+            self.updateCurrentChapter()
+        }
+    }
+    var currentChapter: Chapter?
+    private var cachedChapter: Chapter?
```
Comment (GianniCarlo) [inline Audiobook Player/Models/Book.swift]: what's the use for `cachedChapter` in relation to `currentChapter`?

File: Audiobook Player/Models/Book.swift [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    private func updateCurrentChapter() {
+        // Don't change anything if we have no chapter or are still in range of the current
+        if
```
Comment (GianniCarlo) [inline Audiobook Player/Models/Book.swift]: I'd rather we use `guard` statements where there are only `return`s

**Comment by GianniCarlo** [comment]
The pan gesture to dismiss the player should only be enabled when dragging from the top of the player, right now you can drag from anywhere in the player to dismiss it. This might be debatable actually 🤔, tell me what you whink

File: Audiobook Player/Models/Book.swift (hunk 3 above)
Comment (pichfl) [reply Audiobook Player/Models/Book.swift]: Whelp, that is an artifact from some step of my refactoring. No need for that, will clean up

**Comment by pichfl** [comment]
I would argue that it is actually quite useful, especially on big devices. We do not have any other gesture that could interfere and running this on my personal phone feels quite natural.

File: Audiobook Player/Models/Book.swift (hunk 2 above)
Comment (GianniCarlo) [reply Audiobook Player/Models/Book.swift]: I didn't specify, but I'd rather make `percentCompletedString` a `get` only variable, that way we handle one less state in this class

Comment (GianniCarlo) [reply Audiobook Player/Models/Book.swift]: That `get` would be based on `percentCompleted`

**Comment by GianniCarlo** [comment]
noted, no problem about leaving the gesture as it is. I'd really like to merge this into `next/major`, let me know what other things you consider that are missing to complete this refactor (I don't expect the redesign of the player to be completed in this pr btw). I think this is a solid base for `3.0` and I'd really like to have this as a starting point for my next commits as soon as possible 👍

**Comment by pichfl** [comment]
Will wrap up tonight 🙂

**Comment by GianniCarlo** [comment]
Awesome!

**Pull request merged**